lib/subgraphManagers/IntrospectAddNodeAndCompose.ts

```diff
@@ -0,0 +1,321 @@
+      id: ID! @external
+    }
+  `;
+
+  nodeServiceLocalGraphQLDataSource: LocalGraphQLDataSource;
+
+  private originalSubgraphs: Service[];
+
+  private config: IntrospectAndComposeOptions;
+
+  private nodeGatewayConfig: NodeGatewayConfig;
+
+  private update?: SupergraphSdlUpdateFunction;
+
+  private getDataSource?: GetDataSourceWithSchema;
+
+  private healthCheck?: SubgraphHealthCheckFunction;
+
+  private subgraphs?: (ServiceEndpointDefinition & {
+    dataSource: GraphQLDataSource;
+  })[];
+
+  private serviceSdlCache: Map<string, string> = new Map();
+
+  private timerRef: NodeJS.Timeout | null = null;
+
+  private state: State;
+
+  private completeSupergraphSdl: string;
+
+  constructor(
+    nodeGatewayConfig: NodeGatewayConfig,
+    options: IntrospectAndComposeOptions,
+  ) {
+    this.nodeGatewayConfig = nodeGatewayConfig;
+    this.config = options;
+    this.state = { phase: 'initialized' };
+  }
+
+  public async initialize({
+    update,
+    getDataSource,
+    healthCheck,
+  }: SupergraphSdlHookOptions): Promise<{
+    supergraphSdl: string;
+    cleanup: () => Promise<void>;
+  }> {
+    this.update = update;
+
+    if (this.config.subgraphHealthCheck) {
+      this.healthCheck = healthCheck;
+    }
+    this.getDataSource = getDataSource;
+
+    this.subgraphs = this.config.subgraphs.map(subgraph => ({
+      ...subgraph,
+      dataSource: getDataSource(subgraph),
+    }));
+
+    this.originalSubgraphs = [...this.subgraphs];
+
+    let initialSupergraphSdl: string | null =
+      this.completeSupergraphSdl || null;
+    try {
+      initialSupergraphSdl = await this.updateSupergraphSdl().then(sdl => {
+        if (sdl) {
+          return sdl;
+        }
+        return null;
+      });
+    } catch (e) {
+      this.logUpdateFailure(e);
+      throw e;
+    }
+
+    // Start polling after we resolve the first supergraph
+    if (this.config.pollIntervalInMs) {
+      this.beginPolling();
+    }
+
+    if (!initialSupergraphSdl) {
+      throw new NoInitialSdlError();
+    }
+
+    this.completeSupergraphSdl = initialSupergraphSdl;
+
+    return {
+      cleanup: async (): Promise<void> => {
+        if (this.state.phase === 'polling') {
+          await this.state.pollingPromise;
+        }
+        this.state = { phase: 'stopped' };
+        if (this.timerRef) {
+          clearTimeout(this.timerRef);
+          this.timerRef = null;
+        }
+      },
+      supergraphSdl: initialSupergraphSdl,
+    };
+  }
+
+  private async updateSupergraphSdl(): Promise<string | null> {
+    if (!this.originalSubgraphs.length) {
+      throw new NoSubgraphsError();
+    }
+    const result = await loadServicesFromRemoteEndpoint({
+      getServiceIntrospectionHeaders: async service => {
+        return typeof this.config.introspectionHeaders === 'function'
+          ? // eslint-disable-next-line no-return-await
+            await this.config.introspectionHeaders(service)
```

Comment:
the warning is just that you don't need to `await` since your re in a return... you should remove the eslint disable and avoid the `async`, replacing the constant with `Promise.resolve()`:

```ts
getServiceIntrospectionHeaders: service => {
        return typeof this.config.introspectionHeaders === 'function'
          ? this.config.introspectionHeaders(service)
          : Promise.resolve(this.config.introspectionHeaders);
      },
```